src/snowflake/cli/plugins/stage/diff.py

```diff
@@ -156,109 +167,123 @@ def build_md5_map(list_stage_cursor: DictCursor) -> Dict[str, str]:
-    local_path: Path,
+def preserve_from_diff(
+    diff: DiffResult, stage_paths_to_sync: Collection[StagePath]
+) -> DiffResult:
+    """
+    Returns a filtered version of the provided diff, keeping only the provided stage paths.
+    """
+    preserved_diff: DiffResult = DiffResult()
+    preserved_diff.identical = [i for i in diff.identical if i in stage_paths_to_sync]
+    preserved_diff.different = [i for i in diff.different if i in stage_paths_to_sync]
+    preserved_diff.only_local = [i for i in diff.only_local if i in stage_paths_to_sync]
+    preserved_diff.only_on_stage = [
+        i for i in diff.only_on_stage if i in stage_paths_to_sync
+    ]
+    return preserved_diff
+
+
+def compute_stage_diff(
+    local_root: Path,
     stage_fqn: str,
 ) -> DiffResult:
     """
     Diffs the files in a stage with a local folder.
     """
     stage_manager = StageManager()
-    local_files = enumerate_files(local_path)
+    local_files = enumerate_files(local_root)
     remote_md5 = build_md5_map(stage_manager.list_files(stage_fqn))
 
     result: DiffResult = DiffResult()
 
     for local_file in local_files:
-        relpath = str(local_file.relative_to(local_path))
-        if relpath not in remote_md5:
+        relpath = local_file.relative_to(local_root)
+        stage_filename = to_stage_path(relpath)
+        if stage_filename not in remote_md5:
             # doesn't exist on the stage
-            result.only_local.append(relpath)
+            result.only_local.append(stage_filename)
         else:
             # N.B. we could compare local size vs remote size to skip the relatively-
             # expensive md5sum operation, but after seeing a comment that says the value
             # may not always be correctly populated, we'll ignore that column.
-            stage_md5sum = remote_md5[relpath]
+            stage_md5sum = remote_md5[stage_filename]
             if is_valid_md5sum(stage_md5sum) and stage_md5sum == compute_md5sum(
                 local_file
             ):
                 # the file definitely hasn't changed
-                result.identical.append(relpath)
+                result.identical.append(stage_filename)
             else:
                 # either the file has changed, or we can't tell if it has
-                result.different.append(relpath)
+                result.different.append(stage_filename)
 
             # mark this file as seen
-            del remote_md5[relpath]
+            del remote_md5[stage_filename]
 
     # every entry here is a file we never saw locally
-    for relpath in remote_md5.keys():
-        result.only_on_stage.append(relpath)
+    for stage_filename in remote_md5.keys():
+        result.only_on_stage.append(stage_filename)
 
     return result
 
 
-def get_stage_path_from_file(filepath: str):
-    parent = str(Path(filepath).parent)
-    stage_path = "" if parent == "." else parent
-    return stage_path
+def get_stage_subpath(stage_path: StagePath) -> str:
```

Comment:
Can we call this `get_stage_parent_path` instead?

Reply:
Context: I went with that this is called when used, because it's not quite the parent, it's empty if the parent is ".". Weird contract, I didn't want to go overboard with the refactorings so I kept this method but renamed it to avoid the even more ambiguous name it had before.

However, "." -> "" isn't a particularly big part of the contract, so I can rename as suggested.

Reply:
I'll leave it up to you. I can see both arguments.

Reply:
Ah, now I remember. I was trying to preserve the invariant that anywhere "stage_path" or "stage path" is used, it means a `StagePath` instance in the code. This is not a proper stage path, so I avoided the term on purpose. Will leave as is then, if that's alright with you. I don't want to introduce confusion by making `stage_parent_path` return a string.